### DISCUSSION TITLE
Isotropic Elasticity Tensor as a function of space.

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#18682] [Q&A Modules: Solid mechanics] [answered]
Hi,
I am trying to model 2D elasticity problem where the Poisson's ratio is constant (in space), whereas the Young's modulus is varying. So I used `functions` to define Young's modulus as a function in space and used `GenericFunctionRankTwoTensor` to define the material properties. 
I am getting the following error when trying to run the executable:

> 

> *** ERROR ***
> The following error occurred in the object "MOOSE Problem", of type "FEProblem".
> 
> Material property 'elasticity_tensor', requested by 'stress' is not defined on block 0
> Material property 'elasticity_tensor', requested by 'stress_face' is not defined on block 0
> Material property 'elasticity_tensor', requested by 'stress_neighbor' is not defined on block 0
> the part of the input file that shows the error is as follows: 

```
[Functions]
    [./E]
    type = ParsedFunction
    value = 'x'
    [../]
[]

[AuxVariables]
    [stress_xx]
    order=constant
    family=MONOMIAL
    []
[]

[AuxKernels]
    [aux_stress_xx]
    type=RankTwoScalarAux
    variable=stress_xx
    rank_two_tensor=stress
    execute_on=timestep_end
    scalar_type=AxialStress
    []
[]

[Kernels]
    [TensorMechanics]
    add_variables = true
    displacements = 'dispx dispy'
    []
  
    [body_force]
    type=BodyForce
     variable=dispx
    value=10e3
    []
[]

[Materials]

    [elasticity_tensor]
    type = GenericFunctionRankTwoTensor
    tensor_name = function
    tensor_functions = 'E 0 0 0 0 0 
                        0 E 0 0 0 0
                        0 0 E 0 0 0
                        0 0 0 1 0 0
                        0 0 0 0 1 0
                        0 0 0 0 0 1'
             
[]
```
Any suggestion on fixing this? 

Thanks.

**Comment by jiangwen84** [comment, marked as answer]
1. elasticity_tensor is defined as a RankFourTensor in MOOSE-TensorMechanics
2. Maybe you can try `ComputeVariableIsotropicElasticityTensor`, see https://mooseframework.inl.gov/source/materials/ComputeVariableIsotropicElasticityTensor.html

**Reply by alexchen4ai** [reply]
Is there a similar kernel for transverse isotropic material?

**Reply by hugary1995** [reply]
@AlexCHEN-Engineer I don't think we have one in tensor_mechanics. You can follow `ComputeVariableIsotropicElasticityTensor` to define your own `ComputeVariableOrthotropicElasticityTensor`. In `validParams()`, instead of requesting `youngs_modulus` and `poissons_ratio`, you would request 9 independent material properties.

Alternatively, you can use `DenseMatrix<Real>` instead of `RankFourTensor` to define the orthotropic elasticity tensor to save you some memory, with Hill notation.

In the end, if you are happy with your implementation, and you think it could be useful for other users, you are welcome to submit a pull request!

**Reply by chakra34** [reply]
@hugary1995  @AlexCHEN-Engineer : Is such a temperature dependent Orthotropic elasticity material already a part of MOOSE now ?

**Reply by hugary1995** [reply]
No, I am not aware of any plan to extend the support of temperature-dependent anisotropic elasticity in tensor_mechanics.

In the long term you may consider using https://github.com/idaholab/blackbear. I think you are familiar with deer, and @reverendbedford and I are deprecating deer in favor of blackbear, where we might have more flexibility in supporting your specific needs.

**Reply by chakra34** [reply]
Great! Definitely looking forward to it :) 
Thanks :)
